gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+        // take a reference to it without locking the state. With this, while we are waiting for
+        // the interrupt, set_irq_type() can come and set the state.request to None and we will be
+        // able to quit the wait loop.
+        let request = {
+            let state = &self.state[gpio as usize].write().unwrap();
+
+            match &state.request {
+                Some(x) => x.clone(),
+                None => return Err(Error::GpioIrqNotEnabled),
+            }
+        };
+
+        // Wait for the interrupt for a second.
+        match request.edge_event_wait(Duration::new(1, 0)) {
+            Err(LibGpiodError::OperationTimedOut) => return Err(Error::GpioIrqOpTimedOut),
+            x => x.map_err(Error::GpiodFailed)?,
+        }
+
+        // The interrupt has already occurred, we can lock now just fine.
+        let state = &self.state[gpio as usize].write().unwrap();
+        if let Some(buffer) = &state.buffer {
+            request
+                .edge_event_read(buffer, 1)
+                .map_err(Error::GpiodFailed)?;
+
+            Ok(())
+        } else {
+            Err(Error::GpioLineRequestNotConfigured)
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct LineState {
+    dir: u8,
+    val: Option<u16>,
+    irq_type: u16,
+}
+
+#[derive(Debug)]
+pub(crate) struct GpioController<D: GpioDevice> {
+    config: VirtioGpioConfig,
+    device: D,
+    line_state: Vec<RwLock<LineState>>,
+    line_names: String,
+}
+
+impl<D: GpioDevice> GpioController<D> {
+    // Creates a new controller corresponding to `device`.
+    pub(crate) fn new(device: D) -> Result<GpioController<D>> {
+        let ngpio = device.get_num_lines()? as u16;
+
+        // Allocate enough capacity in the beginning, we can always extend it later.
+        let mut line_names = String::with_capacity((ngpio * 10).into());
+        let mut line_state = Vec::with_capacity(ngpio as usize);
+
+        for offset in 0..ngpio {
+            let name = device.get_line_name(offset)?;
+
+            // Create line names
+            line_names.push_str(&name);
+            line_names.push('\0');
+
+            line_state.push(RwLock::new(LineState {
+                dir: device.get_direction(offset)?,
+                val: None,
+                irq_type: VIRTIO_GPIO_IRQ_TYPE_NONE,
+            }));
+        }
+
+        Ok(GpioController {
+            config: VirtioGpioConfig {
+                ngpio: From::from(ngpio),
+                padding: From::from(0),
+                gpio_names_size: From::from(line_names.len() as u32),
+            },
+            device,
+            line_state,
+            line_names,
+        })
+    }
+
+    pub(crate) fn get_num_lines(&self) -> u16 {
+        self.device.get_num_lines().unwrap()
+    }
+
+    fn get_direction(&self, gpio: u16) -> Result<u8> {
+        Ok(self.line_state[gpio as usize].read().unwrap().dir)
+    }
+
```

Comment:
I'm confused - in get_num_lines() the HW (i.e self.device) is queried but in get_direction() the shadow self.state_line is queried.  Why keeping a shadow state of the device when the device itself can be queried?